SPEC.md

```diff
@@ -490,415 +677,1104 @@ An `Array` represents an ordered list of elements that are all of the same type.
+The [`read_lines`](#read_lines) function presents a special case in which the `Array[String]` value it returns may be immediately coerced into other `Array[P]` values, where `P` is a primitive type. See [Appendix A](#array-deserialization-using-read_lines) for details and best practices.
 
-```wdl
-# simple expressions
-Float f = 1 + 2.2
-Boolean b = if 1 > 2 then true else false
-Map[String, Int] = as_map(zip(["a", "b", "c"], [1, 2, 3]))
+###### Order of Precedence
+
+During string interpolation, there are some operators for which it is possible to coerce the same arguments in multiple different ways. For such operators, it is necessary to define the order of precedence so that a single function prototype can be selected from among the available options for any given set of arguments.
+
+The `+` operator is overloaded for both numeric addition and `String` concatenation. This can lead to the following kinds of situations:
 
-# non-simple expressions
-Int i = x + 3  # requires knowing the value of x
-# requires reading a file that might only exist at runtime
-String s = read_string("/path/to/file")  
+```
+String s = "1.0"
+Float f = 2.0
+String x = "~{s + f}"
 ```
 
-#### Built-in Operators
+There are two possible ways to evaluate the `s + f` expression:
 
-WDL provides the standard unary and binary mathematical and logical operators. The following table lists the valid operand and result type combinations for each operator. Using an operator with unsupported types results in an error.
+1. Coerce `s` to `Float` and perform floating point addition, then coerce to `String` with the result being `x = "3.0"`.
+2. Coerce `f` to `String` and perform string concatenation with result being `x = "1.02.0"`.
 
-In operations on mismatched numeric types (e.g. `Int` + `Float`), the `Int` type is first cast to a `Float`; the result type is always `Float`. This may result in loss of precision, for example if the `Int` is too large to be represented exactly by the `Float`. Note that a `Float` can be converted to an `Int` with the [`ceil`, `round`, or `floor`](#int-floorfloat-int-ceilfloat-and-int-roundfloat) functions.
+Similarly, the equality/inequality operators can be applied to any primitive type.
 
-##### Unary Operators
+The order of precedence is:
 
-| Operator | RHS Type  | Result    |
-| -------- | --------- | --------- |
-| `-`      | `Float`   | `Float`   |
-| `-`      | `Int`     | `Int`     |
-| `!`      | `Boolean` | `Boolean` |
+1. `Int`, `Float`: `Int` coerces to `Float`
+2. `X`, `Y`: For any primitive types `X` != `Y`, both are coerced to `String`
+3. If applying `+` to two values of the same type that cannot otherwise be summed/concatenated (i.e., `Boolean`, `File`, `Directory`), both values are first coerced to `String`
```

Comment:
This stanza is a little ambiguous to me and does not clearly define the order of oeprations. Could we potentially add an example or several examples?